ARROW-16977: [R] Update dataset row counting so no integer overflow on large datasets

thisisnic opened the pull request:
Follow up to #13482 after this one was missed.

nealrichardson reviewed (commented):
Thanks! Two questions:

* I don't see a test of the `r_vec_size()` behavior in this PR or the other one. What is our assurance that it behaves as we want?
* I noticed that this function previously returned `int64_t`, so I wondered whether cpp11 just automatically handled that in the way we wanted. [Turns out someone 🤦 already found out that this doesn't behave nicely], and that we need to handle it ourselves like you have done here. Unfortunately, we have about 20 other functions that return int64_t that would be subject to the same overflow problem. (I searched `^int64_t` in `r/src/` to find them.) Would you mind fixing them all in this PR too?

nealrichardson commented:
Also, it looks like you need to install the package/run the codegen script so that arrowExports.cpp gets updated.

thisisnic commented:
> I don't see a test of the r_vec_size() behavior in this PR or the other one. What is our assurance that it behaves as we want?

I added some in an earlier version of the previous PR, but they took ages to run locally (though were successful), and on CI just crashed it entirely.  I removed them, but couldn't think of an alternative.  Do you have any suggestions?

> ...Would you mind fixing them all in this PR too?

Done.

nealrichardson commented on r/tests/testthat/test-Table.R:
```diff
@@ -696,3 +696,18 @@ test_that("as_arrow_table() errors for invalid input", {
   )
 })
+
+test_that("num_rows method not susceptible to integer overflow", {
```
These large memory tests are slow, so I don't think you need two versions, I think you can keep just the Table version. 

If you want to test more places r_vec_size is used, you can get `$nbytes()` on the arrays/chunkedarrays, as well as `length()`.

thisisnic replied:
I've removed the RecordBatch test, added in tests for a few more fields, as well as updating a couple of utility functions that these changes broke.  The one thing I didn't do is getting `array$data()` and actually, I'm wondering if overflow is possible with that?  From my understanding, an array's maximum size would be the maximum integer value, or am I getting mixed up there?

nealrichardson replied:
On the format, see https://github.com/apache/arrow/blob/master/format/Schema.fbs#L155-L171 and https://arrow.apache.org/docs/format/Columnar.html. 

In R:

```
# Here's a simple integer array
> a <- Array$create(1:5)
> a$data()
ArrayData
> a$data()$buffers
[[1]]
NULL

[[2]]
Buffer
# the first buffer is for the null bitmask, but it's empty because there are no NAs

> a$data()$buffers[[2]]
Buffer
> a$data()$buffers[[2]]$size
[1] 20
> a$data()$buffers[[2]]$capacity
[1] 20

# here's a string array
> strings <- Array$create(letters)
# note it has 3 buffers, one for the data and one for the offsets
> strings$data()$buffers
[[1]]
NULL

[[2]]
Buffer

[[3]]
Buffer

> strings$data()$buffers[[2]]$size
[1] 108
> strings$data()$buffers[[3]]$size
[1] 26

# let's make a LargeString array using the helper from the test suite
> make_big_string <- function() {
+   # This creates a character vector that would exceed the capacity of BinaryArray
+   rep(purrr::map_chr(2047:2050, ~ paste(sample(letters, ., replace = TRUE), collapse = "")), 2^18)
+ }
> big <- Array$create(make_big_string())
> big$type
LargeUtf8
large_string
> big$data()$buffers
[[1]]
NULL

[[2]]
Buffer

[[3]]
Buffer

> big$data()$buffers[[2]]$size
[1] 8388616
# the data buffer is > MAX_INT32 so it overflows (on master)
> big$data()$buffers[[3]]$size
[1] -2146959360
```
